src/metadata/metadata_superblock.c

```diff
@@ -98,42 +98,135 @@ static void ocf_metadata_store_segment(ocf_pipeline_t pipeline,
+		return -OCF_ERR_INVAL;
+	}
+
+	if (superblock->clean_shutdown > ocf_metadata_clean_shutdown) {
+		ocf_log_invalid_superblock("shutdown status");
+		return -OCF_ERR_INVAL;
+	}
+
+	if (superblock->dirty_flushed > DIRTY_FLUSHED) {
+		ocf_log_invalid_superblock("flush status");
+		return -OCF_ERR_INVAL;
+	}
+
+	if (superblock->flapping_idx > 1) {
+		ocf_log_invalid_superblock("flapping index");
+		return -OCF_ERR_INVAL;
+	}
+
+	if (superblock->cache_mode < 0 ||
+			superblock->cache_mode >= ocf_cache_mode_max) {
+		ocf_log_invalid_superblock("cache mode");
+		return -OCF_ERR_INVAL;
+	}
+
+	if (env_strnlen(superblock->name, sizeof(superblock->name)) >=
+			OCF_CACHE_NAME_SIZE) {
+		ocf_log_invalid_superblock("name");
+		return -OCF_ERR_INVAL;
+	}
+
+	if (superblock->valid_parts_no > OCF_USER_IO_CLASS_MAX) {
+		ocf_log_invalid_superblock("partition count");
+		return -OCF_ERR_INVAL;
+	}
+
+	if (!ocf_cache_line_size_is_valid(superblock->line_size)) {
+		ocf_log_invalid_superblock("cache line size");
+		return -OCF_ERR_INVAL;
+	}
+
+	if ((unsigned)superblock->metadata_layout >= ocf_metadata_layout_max) {
+		ocf_log_invalid_superblock("metadata layout");
+		return -OCF_ERR_INVAL;
+	}
+
+	if (superblock->core_count > OCF_CORE_MAX) {
+		ocf_log_invalid_superblock("core count");
+		return -OCF_ERR_INVAL;
+	}
+
+	if (superblock->cleaning_policy_type < 0 ||
+			superblock->cleaning_policy_type >= ocf_cleaning_max) {
+		ocf_log_invalid_superblock("cleaning policy");
+		return -OCF_ERR_INVAL;
+	}
+
+	if (superblock->promotion_policy_type < 0 ||
+			superblock->promotion_policy_type >=
+					ocf_promotion_max) {
+		ocf_log_invalid_superblock("promotion policy");
+		return -OCF_ERR_INVAL;
 	}
 
+	return 0;
+}
+
+static void _ocf_metadata_validate_superblock(ocf_pipeline_t pipeline,
+		void *priv, ocf_pipeline_arg_t arg)
+{
+	struct ocf_metadata_context *context = priv;
+	struct ocf_metadata_ctrl *ctrl;
+	struct ocf_superblock_config *superblock;
+	int ret;
+
+	ctrl = (struct ocf_metadata_ctrl *)context->ctrl;
+	superblock = METADATA_MEM_POOL(ctrl, metadata_segment_sb_config);
+
+	ret = ocf_metadata_validate_superblock(NULL, superblock); // ocf_ctx_t here - how?
+	if (ret)
+		OCF_PL_FINISH_RET(pipeline, ret);
+
 	ocf_pipeline_next(pipeline);
 }
 
 static void ocf_metadata_load_superblock_post(ocf_pipeline_t pipeline,
 		void *priv, ocf_pipeline_arg_t arg)
 {
 	struct ocf_metadata_context *context = priv;
+	ocf_cache_t cache = context->cache;
 	struct ocf_metadata_ctrl *ctrl;
 	struct ocf_superblock_config *sb_config;
-	ocf_cache_t cache = context->cache;
 
-	ctrl = (struct ocf_metadata_ctrl *)cache->metadata.priv;
+	struct ocf_metadata_uuid *muuid;
+	struct ocf_volume_uuid uuid;
+	ocf_volume_type_t volume_type;
+	ocf_core_t core;
+	ocf_core_id_t core_id;
+
+
+	ctrl = (struct ocf_metadata_ctrl *)context->ctrl;
 	sb_config = METADATA_MEM_POOL(ctrl, metadata_segment_sb_config);
 
 	if (sb_config->core_count > OCF_CORE_MAX) {
```

Comment:
these checks (lines 232-242) are now included in ocf_metadata_validate_superblock. If I recall corectly they were removed from here in 21.3 / 21.6 patches